packages/transaction-controller/src/utils/simulation.ts

```diff
@@ -477,24 +477,47 @@ function getEventTokenIds(event: ParsedEvent): (Hex | undefined)[] {
 }
 
 /**
- * Extract the value from a balance transaction response.
+ * Extract the value from a balance transaction response using the correct ABI.
  * @param from - The address to check the balance of.
  * @param token - The token to check the balance of.
  * @param response - The balance transaction response.
- * @returns The value of the balance transaction.
+ * @returns The value of the balance transaction as Hex.
  */
-function getValueFromBalanceTransaction(
+export function getValueFromBalanceTransaction(
   from: Hex,
   token: SimulationToken,
   response: SimulationResponseTransaction,
 ): Hex {
-  const normalizedReturn = normalizeReturnValue(response.return);
+  const interfaces = getContractInterfaces();
+  let decoded;
 
-  if (token.standard === SimulationTokenStandard.erc721) {
-    return normalizedReturn === from ? '0x1' : '0x0';
+  try {
+    if (token.standard === SimulationTokenStandard.erc721) {
+      // ERC721: Check ownership and return 1 if the owner is the user, 0 otherwise.
+      decoded = (
```

Comment:
Minor, since the `SupportedToken` and `SimulationTokenStandard` values overlap, can we remove some of the duplication?

```
const contract = interfaces.get(token.standard as SupportedToken) as Interface;
const isERC721 = token.standard === SimulationTokenStandard.erc721;
const functionName = isERC721 ? 'ownerOf' : 'balanceOf';
const functionResult = contract.decodeFunctionResult(functionName, response.return);
const value = functionResult[0].toLowerCase();

return isERC721 ?
    value === from.toLowerCase() ? '0x1' : '0x0' :
    value;
```

Reply:
I'm afraid I don't follow this logic, I feel it's clearer and more readable the less duplication we have, and therefore the fewer lines of code the reader has to parse.

We still have explicit separate paths for ERC-721 except they're encapsulated via a boolean so the differences in that path are more readable as we can quickly see it impacts the `functionName` and returned value, but otherwise the logic is the same.

Reply:
Hey Matt, I have since extracted the interface creation to keep each branch shorter. I still think combining them doesn't quite make sense for clarity, even more-so now that they are just a couple lines each.

Also, I wanted to let you know that the ownership comparison change fixes a bug – we were trimming leading zeros from addresses, so those starting with zeros weren’t matching the 'from' address and missed ERC721 balance updates. I added a regression test to cover this.